### PR TITLE
Refactor: Add Response Handler

### LINKED ISSUE #648: Use a response handler to convert Response to HttpResponse
**What is not easy to do right now?**
The `Handler.scala` file contains a lot of code to handle `Response` and `HttpData`, which is hard to maintain and refactor.

**Describe the solution you'd like**
Create a new ChannelHandler that can be added at the time of channel initialization, to handle the response types on the Server.

**Why is it important?**
It will make the code inside of `Handler.scala` much easier to refactor.

**Additional context**
Some similar work is done in the client #636

### PR DESCRIPTION
**Pull request by gciuloaica**
…io-http Response in netty HTTP response.

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/service/Handler.scala [hunk 1]
```diff
@@ -156,22 +95,18 @@ private[zhttp] final case class Handler[R](
...
-            case _                             => releaseRequest(jReq)
-          }
+          ctx.fireChannelRead(res)
+          releaseRequest(jReq)
         }
-      case HExit.Failure(e)   =>
-        unsafeWriteAndFlushErrorResponse(e)
+
+      case HExit.Failure(e) =>
+        ctx.fireExceptionCaught(e)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/Handler.scala]:
```suggestion
        ctx.fireChannelRead(e)
```

It's not an error in this context. We have an 500 status code response which is essentially a success value.

File: zio-http/src/main/scala/zhttp/service/server/ServerChannelInitializer.scala [hunk 2]
```diff
@@ -61,6 +62,9 @@ final case class ServerChannelInitializer[R](
     // Always add ZIO Http Request Handler
     pipeline.addLast(HTTP_REQUEST_HANDLER, reqHandler)
 
+    // ServerResponseHandler - transforms intern Response in HttpResponse
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/server/ServerChannelInitializer.scala]:
```suggestion
    // ServerResponseHandler - transforms Response to HttpResponse
```

File: zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  config: Server.Config[R, Throwable],
+  serverTime: ServerTimeGenerator,
+) extends SimpleChannelInboundHandler[Response](false) {
+
+  type Ctx = ChannelHandlerContext
+
+  override def channelRead0(ctx: Ctx, response: Response): Unit = {
+    implicit val iCtx: ChannelHandlerContext = ctx
+
+    response.status match {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala]: why do we need to check for `Status`? Can't we simply encode and write to the channel, like we were doing before?

File: zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    }
+    ()
+  }
+
+  override def exceptionCaught(ctx: Ctx, cause: Throwable): Unit = {
+    if (ctx.channel().isWritable) {
+      ctx.writeAndFlush(serverErrorResponse(cause)): Unit
+    }
+    config.error.fold(super.exceptionCaught(ctx, cause))(f => runtime.unsafeRun(ctx)(f(cause)))
+  }
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala:38]: this should only handle internal errors. Basically for logging etc.

File: zio-http/src/main/scala/zhttp/service/Handler.scala (hunk 1 above)
Comment (gciuloaica) [reply zio-http/src/main/scala/zhttp/service/Handler.scala]: right. updated code

File: zio-http/src/main/scala/zhttp/service/server/ServerChannelInitializer.scala (hunk 2 above)
Comment (gciuloaica) [reply zio-http/src/main/scala/zhttp/service/server/ServerChannelInitializer.scala]: updated

File: zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala (hunk 3 above)
Comment (gciuloaica) [reply zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala]: simplified the code by avoiding pattern matching on status

File: zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala (hunk 4 above)
Comment (gciuloaica) [reply zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala:38]: done

File: zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    ctx.write(encodeResponse(response))
+    response.data match {
+      case HttpData.BinaryStream(stream) =>
+        runtime.unsafeRun(ctx) {
+          writeStreamContent(stream)
+        }
+      case HttpData.File(file)           =>
+        unsafeWriteFileContent(file)
+      case _                             => ctx.flush()
+    }
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala]:
```suggestion
      case HttpData.BinaryStream(stream) => runtime.unsafeRun(ctx) { writeStreamContent(stream) }
      case HttpData.File(file)           => unsafeWriteFileContent(file)
      case _                             => ctx.flush()
    }
```

Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala]: style nitpick :)

File: zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+      // Check if the encoded response exists and/or was modified.
+      case Some((oRes, jResponse)) if oRes eq res =>
+        jResponse match {
+          // Duplicate the response without allocating much memory
+          case response: FullHttpResponse =>
+            response.retainedDuplicate()
+
+          case response =>
+            response
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/service/server/content/handlers/ServerResponseHandler.scala]:
```suggestion
          case response: FullHttpResponse => response.retainedDuplicate()
          case response                   => response
```

**Review by tusharmath — APPROVED** [review]

**Pull request merged**